include/envoy/ssl/handshaker.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/ssl/socket_state.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Ssl {
+
+class HandshakerCallbacks {
+public:
+  virtual ~HandshakerCallbacks() = default;
+
+  /**
+   * Hands off the internally-held SSL object for external manipulation.
+   */
+  virtual bssl::UniquePtr<SSL> Handoff() PURE;
+  /**
+   * Accepts an SSL object for internal storage.
+   */
+  virtual void Handback(bssl::UniquePtr<SSL> ssl) PURE;
```

Comment:
Do we actually need those 2 functions or are we leaking Split Handshake implementation into the API?

Reply:
Agreed, this is a bit leaky. But I don't see another way to expose these methods to private implementations of Handshaker, WDYT?

Reply:
You got the same kind of feedback from me on the internal review.  Please consider having the "handshaker" own the SSL* even after handshake completes and either provide an SSL* accessor or fully wrap all SSL functions including SSL_read/SSL_write.

Reply:
This suggestion helped me a lot; it let me both (a) remove the Handback()/Handoff() methods from this Callbacks struct and (b) remove the SSL* argument from the general Handshaker interface. 👍